homeassistant/components/mqtt/device_trigger.py

```diff
@@ -0,0 +1,210 @@
+        if discovery_hash in triggers:
+            triggers[discovery_hash] = Trigger(
+                device_id=device_id,
+                type=config[CONF_TYPE],
+                subtype=config[CONF_SUBTYPE],
+                topic=config[CONF_TOPIC],
+                payload=config[CONF_PAYLOAD],
+                qos=config[CONF_QOS],
+                discovery_hash=discovery_hash,
+            )
+
+
+def _remove_trigger(hass, discovery_hash):
+    for _, triggers in hass.data[DEVICE_TRIGGERS].items():
+        triggers.pop(discovery_hash)
+
+
+async def async_setup_trigger(hass, config, config_entry, discovery_hash):
+    """Set up the MQTT device trigger."""
+    config = PLATFORM_SCHEMA(config)
+    remove_signal = None
+
+    async def discovery_update(payload):
+        """Handle discovery update."""
+        _LOGGER.info(
+            "Got update for trigger with hash: %s '%s'", discovery_hash, payload
+        )
+        if not payload:
+            # Empty payload: Remove trigger
+            _LOGGER.info("Removing trigger: %s", discovery_hash)
+            _remove_trigger(hass, discovery_hash)
+            clear_discovery_hash(hass, discovery_hash)
+            remove_signal()
+        else:
+            # Non-empty payload: Update trigger
+            _LOGGER.info("Updating trigger: %s", discovery_hash)
+            payload.pop(ATTR_DISCOVERY_HASH)
+            config = PLATFORM_SCHEMA(payload)
+            await _update_device(hass, config_entry, config)
+            _update_trigger(hass, discovery_hash, config)
+
+    remove_signal = async_dispatcher_connect(
+        hass, MQTT_DISCOVERY_UPDATED.format(discovery_hash), discovery_update
+    )
+
+    await _update_device(hass, config_entry, config)
+
+    device_registry = await hass.helpers.device_registry.async_get_registry()
+    device = device_registry.async_get_device(
+        {(DOMAIN, id_) for id_ in config[CONF_DEVICE][CONF_IDENTIFIERS]},
+        {tuple(x) for x in config[CONF_DEVICE][CONF_CONNECTIONS]},
+    )
+
+    if device is None:
+        return
+
+    if DEVICE_TRIGGERS not in hass.data:
+        hass.data[DEVICE_TRIGGERS] = {}
+    if device.id not in hass.data[DEVICE_TRIGGERS]:
+        hass.data[DEVICE_TRIGGERS][device.id] = {}
+    hass.data[DEVICE_TRIGGERS][device.id][discovery_hash] = Trigger(
+        device_id=device.id,
+        type=config[CONF_TYPE],
+        subtype=config[CONF_SUBTYPE],
+        topic=config[CONF_TOPIC],
+        payload=config[CONF_PAYLOAD],
+        qos=config[CONF_QOS],
+        discovery_hash=discovery_hash,
+    )
+
+
+async def async_get_triggers(hass: HomeAssistant, device_id: str) -> List[dict]:
+    """List device triggers for MQTT devices."""
+    triggers = []
+
+    if DEVICE_TRIGGERS not in hass.data or device_id not in hass.data[DEVICE_TRIGGERS]:
+        return triggers
+
+    for _, auto in hass.data[DEVICE_TRIGGERS][device_id].items():
+        trigger = dict(MQTT_TRIGGER)
```

Comment:
```python
trigger = {
  **MQTT_TRIGGER,
  'device_id': device_id,
  'type': auto.type
}
```

Reply:
Also `MQTT_TRIGGER` should be named differently, as it is used as `MQTT_TRIGGER_BASE` ?.

Reply:
Fixed.